com.reprezen.swagedit.openapi3/src/com/reprezen/swagedit/openapi3/validation/OpenApi3Validator.java

```diff
@@ -86,4 +88,23 @@ protected void validateOperationIdReferences(Model model, AbstractNode node, Set
         }
     }
 
+    protected void validateParameters(Model model, AbstractNode node, Set<SwaggerError> errors) {
+        final JsonPointer pointer = JsonPointer.compile("/definitions/parameterOrReference");
+
+        if (node != null && node.getType() != null && pointer.equals(node.getType().getPointer())) {
+            // validation parameter location value
+            if (node.isObject() && node.asObject().get("in") != null) {
+                AbstractNode valueNode = node.asObject().get("in");
+                try {
+                    Object value = valueNode.asValue().getValue();
+
+                    if (!Lists.newArrayList("query", "header", "path", "cookie").contains(value)) {
```

Comment:
@ghillairet , @tfesenko , It looks like we did this using special validation logic.  Are we generally trying to avoid making edits to the generated schema? Or have we already diverged from the generated schema? 

I'm wondering if it would be more efficient, generally, to either make these edits directly in the schema (in this case by adding an enum constraint), or post-process the generated schema to add missing details that could not be extracted by markdown-scraping.

Reply:
@tedepstein We did that because the v3 schema does only contain a string type for this value. The v2 did contain an enum type.

Reply:
Yes, the v2 schema was hand-crafted by someone.  The v3 schema is generated by analyzing the OpenAPI 3.0 specification's source markdown, and it misses many of these constraints. 

So my question is really about how we add those missing constraints, as we discover them. I see three general approaches:

    1 Add special-case validation and content assist code, as you've done here. 
    2 Directly edit the generated v3 schema (Which means we won't be able to use updates from a re-generated schema, if there is one. But that might be OK.)
    3 Write some code that modifies the generated v3 schema, and run that code either at design time, or as part of our build, or at runtime. This is a little bit more difficult than  option 2, but would avoid the special-case logic of option 1, and still let us use future versions of the generated v3 schema.